### ISSUE TITLE
fastcore.utils.Tuple is not a replacement for typing.Tuple

### ISSUE THREAD

**Issue by indigoviolet** [fastai/fastcore#37]
Sorry to open a duplicate issue, but I posted (https://github.com/fastai/fastcore/issues/36#issuecomment-628178350) in #36 after it was closed about how fastcore.utils.Tuple is actually not usable in place of typing.Tuple, unlike what @tcappelle said, but perhaps you don't receive notifications from closed issues.

Copying that comment here:

>  Hi, thanks for the quick responses.
> 
> 1. I don't think I can use `fastcore.utils.Tuple` as `typing.Tuple`: see this gist where it breaks: https://gist.github.com/indigoviolet/175556625b9e21748fe2d2be6500a1bb
> 2. I understand that I can `import typing.Tuple as OldTuple`, but IMHO a library shouldn't shadow a standard library name without a very good reason, and should definitely not break it. I would expect `fastcore.utils.Tuple` to require explicitly importing it if it needs to be named `Tuple`.

**Comment by jph00** [comment]
Can you explain more about what you mean here? I don't see how `fastcore.utils.Tuple` "shadows" `typing.Tuple`. They have the same class name, but they are in different modules. `typing.Tuple` is still available as `typing.Tuple`, so it's not shadowed, AFAICT. You can still `from typing import Tuple` if you wish.

`fastcore.utils.Tuple` is not intended to be a replacement for `typing.Tuple`.

IIUC, what you're really saying is that fastai shouldn't name a class with the same name as a class in stdlib. I'm not sure if there's really a justification for that rule, in general.

I'm quite possibly missing something though, so please do let us know more about what specific concerns you have.

**Comment by indigoviolet** [comment]
@jph00, thanks for the response. I should have clarified this better:

```
In [1]:
from fastai2.basics import *

In [2]:
Tuple
Out [2]:
fastcore.utils.Tuple
```
> IIUC, what you're really saying is that fastai shouldn't name a class with the same name as a class in stdlib. I'm not sure if there's really a justification for that rule, in general.

I agree that `typing.Tuple` is not automatically shadowed, and that it is easy to import `typing.Tuple` explicitly. But this kind of error is surprising, it could be avoided. 

I'm suggesting that since fastai encourages the `import *` style, it should be extra careful about default exporting names that conflict with stdlib (or other common data science libraries). 

(Just as a separate data point, I've been bitten similarly recently where I had a function named `one_hot` and then something changed -- either I did another `import *` or pulled master, or I ran cells in a different order than before -- but fastai2's `one_hot` was shadowing mine and it took some time to figure that out. I realize that this is a natural risk of `import *` that you've probably considerd, that fastai can't avoid naming things like `one_hot`, but it makes for a slightly more frustrating developer experience.)
